lib/types: fix either lazy value regression in case of non-matching v…

hsjobeki opened the pull request:
Fixes https://github.com/NixOS/nixpkgs/issues/438459 a regression introduced by #391544

cc @eclairevoyant feel free to test and approve if this solves your problem.

Example of the warning

```nix-repl
nix-repl> nixos.config.services.printing.cups-pdf.instances.pdf.settings
evaluation warning: See https://github.com/NixOS/nixpkgs/issues/438825. If you are the maintainer of this module please fix the followin issue:
                    The option `services.printing.cups-pdf.instances.pdf.settings` is neither a value of type `signed integer or string or absolute path` nor `package`, Definition values:
                    - In `example.nix':
                        {
                          UserUMask = "0033";
                        }
{
  AnonDirName = "/var/spool/cups-pdf-pdf/anonymous";
  Anonuser = "root";
  GhostScript = "/nix/store/lin169x3s56razcp2j6z31nic7h1m07s-ghostscript-with-X-10.05.1/bin/gs";
  Out = "/var/spool/cups-pdf-pdf/users/\${USER}";
  Spool = "/var/spool/cups-pdf-pdf/spool";
  UserUMask = "0033";
}
```

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Yarny0 reviewed (approved):
The pull request at hand indeed fixes https://github.com/NixOS/nixpkgs/issues/438459 .  I tested with https://github.com/NixOS/nixpkgs/issues/438459#issuecomment-3239158059 .

eclairevoyant commented:
Yes this fixes the issue.

But I don't understand why the value in the associated issue was considered "non-matching"?
Doesn't it `UserUMask = "0033";` match `str`?

https://github.com/NixOS/nixpkgs/blob/f8f76dc024047ea8392a95d4f8ec6aa2eb57cbe8/nixos/modules/services/printing/cups-pdf.nix#L43-L50

Yarny0 commented:
@eclairevoyant wrote:
> EDIT: I missed that the `attrsOf` was missing. Actual fix for this module in #438619.

Thanks for that hint.  While I don't really understand the freeform system, from looking at https://github.com/NixOS/nixpkgs/blob/dfb2f12e899db4876308eba6d93455ab7da304cd/nixos/doc/manual/development/freeform-modules.section.md , this is likely the real problem here, and I support fixing this (instead of relaxing the option merge code).

@eclairevoyant wrote:
> I'd prefer to merge the actual fix rather than hide the type error.... but I hope you agree that the error message is not intuitive to read.

Maybe an expansion of the error message along the following diff might help (only superficially tested):

```diff
diff --git i/lib/types.nix w/lib/types.nix
index 6b51f9254a00..f3fcdcf8f106 100644
--- i/lib/types.nix
+++ w/lib/types.nix
@@ -1453,7 +1453,7 @@ let
                       headError = {
                         message = "The option `${showOption loc}` is neither a value of type `${t1.description}` nor `${t2.description}`, Definition values: ${showDefs defs}";
                       };
-                      value = abort "(t.merge.v2 defs).value must only be accessed when `.headError == null`. This is a bug in code that consumes a module system type.";
+                      value = abort "(t.merge.v2 defs).value must only be accessed when `.headError == null`. This is a bug in code that consumes a module system type.  Original error: \n${headError.message}";
                     };
               in
               checkedAndMerged;
```

eclairevoyant commented:
> I don't really understand the freeform system

This doesn't really have anything to do with being a free form type. The type is "attrsOf (etc.)". 

The submodule/freeformType aspect just means you don't need to list out the _inner_ options exhaustively, but you want to specify _some_ of them via `mkOption` as hints/typechecks for the end user.

eclairevoyant commented:
I don't know what headError is but is there a way to emit that?

hsjobeki commented:
cc @MattSturgeon since you are familiar with the history now, could you review this change? I added a warning, though i am not sure about the exact wording.

MattSturgeon reviewed (approved):
LGTM.

I'd be great to have a test case for the fixed regression. Even if it is a deprecated/broken/buggy behaviour, a test would mean we notice if it unintentionally regresses again before the deprecation period is over.

I won't block on this, if you don't think it's worth the effort.

roberth reviewed (changes requested):
> check = x: t1.check x || t2.check x;

&mdash; old nixpkgs `types.either`

The root cause is still the freeform logic continuing despite `headError != null`, right?
Could we move this check to evalModules/freeform? Then we can 
- know which submodule is wrong and report a better message
- not weaken and confuse other uses of either
- maybe emit fewer messages (one per submodule instead of per freeform definition, but also I haven't verified)

I suppose we may need something like `valueMeta.deprecatedUncheckedValue` in order to extract the merged value, and then freeform can fall back to that when `headError != null`.

> Unfortunately some could argue that this behavior has become a "feature".


I suppose other callers of the type interface might have a similar problem, but let's at least start by considering that a bug on their end. Merge without check should always have been considered "undefined behavior" IMO.
What we should do though, is document this problem in the release notes as a breaking change.

hsjobeki commented:
> The root cause is still the freeform logic continuing despite `headError != null`, right?
> Could we move this check to evalModules/freeform? Then we can

Okay i'll summarize the TODOs for a proper fix:

1. Calling `declaredConfig._module.freeformType.merge prefix defs`
Should be changed to:

```nix
type.merge.v2 { ... }
```

2. We should fix the way we implemented backwards compat. calling v2 and returning .value might be subideal.
3. We should also migrate nullOr to use merge.v2 seems I missed that it was a composed type in the original PR.

Should we merge this PR until the proper fix is done?

hsjobeki committed:
lib/types: check headError in merge v1

types.either needs special handling for backwards compatibility
Previous wrong usages in freeformType silently accepted any value

hsjobeki committed:
lib/types: add deprecation note about freeform behavior of 'either'

hsjobeki commented:
@roberth I found that just doing 2. should be sufficient. 

Doing something like `if merge ? v2 then declaredConfig._module.freeformType.merge.v2 { loc defs } else ...` would still use merge.v1 in most cases. because `nullOr`, and other types have not been migrated yet.

hsjobeki committed:
lib/types: add test for regression of either

For details see github.com/NixOS/nixpkgs/issues/438459

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/breaking-changes-announcement-for-unstable/17574/97

hsjobeki commented:
Alternative fix in. https://github.com/NixOS/nixpkgs/pull/440459
